src/mlpack/core/tree/rectangle_tree/hilbert_r_tree_auxiliary_information_impl.hpp

```diff
@@ -0,0 +1,192 @@
+/**
+ * @file hilbert_r_tree_auxiliary_information.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Implementation of the HilbertRTreeAuxiliaryInformation class,
+ * a class that provides some Hilbert r-tree specific information
+ * about the nodes.
+ */
+#ifndef MLPACK_CORE_TREE_RECTANGLE_TREE_HILBERT_R_TREE_AUXILIARY_INFORMATION_IMPL_HPP
+#define MLPACK_CORE_TREE_RECTANGLE_TREE_HILBERT_R_TREE_AUXILIARY_INFORMATION_IMPL_HPP
+
+#include "hilbert_r_tree_auxiliary_information.hpp"
+
+namespace mlpack {
+namespace tree {
+
+
+template<typename TreeType,
+         template<typename> class HilbertValueType>
+HilbertRTreeAuxiliaryInformation<TreeType, HilbertValueType>::
+HilbertRTreeAuxiliaryInformation()
+{
+
+};
+
+template<typename TreeType,
+         template<typename> class HilbertValueType>
+HilbertRTreeAuxiliaryInformation<TreeType, HilbertValueType>::
+HilbertRTreeAuxiliaryInformation(const TreeType* node) :
+    hilbertValue(node)
+{
+
+};
+
+template<typename TreeType,
+         template<typename> class HilbertValueType>
+HilbertRTreeAuxiliaryInformation<TreeType, HilbertValueType>::
+HilbertRTreeAuxiliaryInformation(const HilbertRTreeAuxiliaryInformation& other) :
+    hilbertValue(other.HilbertValue())
+{
+
+};
+
+template<typename TreeType,
+         template<typename> class HilbertValueType>
+HilbertRTreeAuxiliaryInformation<TreeType, HilbertValueType>::
+~HilbertRTreeAuxiliaryInformation()
+{
+
+}
+  
+template<typename TreeType,
+         template<typename> class HilbertValueType>
+bool HilbertRTreeAuxiliaryInformation<TreeType, HilbertValueType>::
+HandlePointInsertion(TreeType* node, const size_t point)
+{
+  if (node->IsLeaf())
+  {
+    // Get the position at which the point should be inserted
+    // Update the largest Hilbert value of the node
+    size_t pos = hilbertValue.InsertPoint(node, node->Dataset().col(point));
+
+    // Move points
+    for (size_t i = node->NumPoints(); i > pos; i--)
+      node->Points()[i] = node->Points()[i-1];
+    // Insert the point
+    node->Points()[pos] = point;
+    node->Count()++;
+  }
+  else
+  {
+    // Calculate the Hilbert value
+    hilbertValue.InsertPoint(node, node->Dataset().col(point));
```

Comment:
I think it would be possible here to just pass `point` instead of `node->Dataset().col(point)`, and then refactor `DiscreteHilbertValue::InsertPoint()` to take a `size_t` instead, since we only have one dataset.  But it is up to you if you want to do that.  I don't think there would be any noticeable advantage or disadvantage.